pkgs/build-support/gradle/update-deps.nix

```
@@ -0,0 +1,158 @@
+    trap "rm -rf '$SRC'" SIGINT SIGTERM ERR EXIT
+    cp -r "$source/src" "$SRC/"
+    chmod -R +w "$SRC/src"
+    cd "$SRC/src"
+    source "$source/script"
+    ${preBuild}
+    ${gradleCommands}
+    ${postBuild}
+  '';
+in
+writeShellScript "fetch-deps.sh" ''
+  set -eo pipefail
+  ${lib.optionalString (lib.hasInfix "$attrPath" code) ''
+    attrPath=${lib.escapeShellArg attrPath}
+    [[ -n "$attrPath" ]] || attrPath="$1"
+    [[ -n "$attrPath" ]] || attrPath="$UPDATE_NIX_ATTR_PATH"
+    [[ -n "$attrPath" ]] || (echo "Set UPDATE_NIX_ATTR_PATH or pass attr path as the first argument!" && exit 1)
+  ''}
+  outPath="${
+    # if this is an absolute path in nix store, use path relative to the output root
+    if lib.hasPrefix builtins.storeDir (toString depsPath)
+    then builtins.concatStringsSep "/" (lib.drop 4 (lib.splitString "/" (toString depsPath)))
+    # if this is an absolute path, just use that path
+    else if lib.hasPrefix "/" (toString depsPath)
+    then depsPath
+    # otherwise, use a path relative to the package
+    else ''$(nix-instantiate --eval -E "dirOf (${code}).meta.position" | tr -d '"')/${depsPath}''
+  }"
+
+  source="$(nix-build --no-out-link -E "import ${./patched-source.nix} ({ self = ${code}; } // builtins.fromJSON "${
+    lib.escapeShellArg (builtins.toJSON (builtins.toJSON attrOverrides))
+  }")")"
+
+  pushd "$(mktemp -d)"
+  MITM_CACHE_DIR="$PWD"
+  trap "rm -rf '$MITM_CACHE_DIR'" SIGINT SIGTERM ERR EXIT
+  ${openssl}/bin/openssl genrsa -out ca.key 2048
+  ${openssl}/bin/openssl req -x509 -new -nodes -key ca.key -sha256 -days 1 -out ca.cer -subj "/C=AL/ST=a/L=a/O=a/OU=a/CN=example.org"
+  export MITM_CACHE_PORT="''${mitmCachePort:-${mitmCachePortBash}}"
+  # forget all redirects - this makes the lockfiles predictable
+  # not only does this strip CDN URLs, but it also improves security - since the redirects aren't
+  # stored in the lockfile, a malicious actor can't change the redirect URL stored in the lockfile
+  ${mitm-cache}/bin/mitm-cache \
+    -l"127.0.0.1:$MITM_CACHE_PORT" \
+    record \
+    --reject '\.(md5|sha(1|256|512:?):?)$' \
+    --forget-redirects-from '.*' \
+    --record-text '/maven-metadata\.xml$' >/dev/null 2>/dev/null &
+  MITM_CACHE_PID="$!"
+  trap "kill '$MITM_CACHE_PID'" SIGINT SIGTERM ERR EXIT
+  sleep 0.5
+  ps -p "$MITM_CACHE_PID" >/dev/null || (echo "Failed to launch mitm-cache! Port $MITM_CACHE_PORT may be already used. Try setting "'$mitmCachePort' && exit 1)
+  export MITM_CACHE_CA="$PWD/ca.cer"
+  export MITM_CACHE_KS_PWD="$(head -c10 /dev/random | base32)"
+  echo y | ${jdk}/bin/keytool -importcert -file ca.cer -alias alias -keystore keystore -storepass "$MITM_CACHE_KS_PWD"
+  export MITM_CACHE_KEYSTORE="$PWD/keystore"
+  popd
+  ${preBuildNoSandbox}
+  # nsjail isn't necessary, it's only used to prevent messy build scripts from touching ~
+  ${nsjail}/bin/nsjail \
+    -N --keep_caps --chroot / --disable_rlimits \
+    -T /home -T /root -R ${path} \
```

Adding `/tmp` as a rw bindmount fixes the issue where it needs to be mounted as tmpfs. Tried to mount it with `-T`, but it didn't work.

```suggestion
    -T /home -T /root -B /tmp -R ${path} \
```

---

I guess `nsjail` keeps the mounts as-is. It's not ideal as the Gradle processes may still e.g. affect something outside of /home and /root as long as it's a mount, but whatever, it'll do.

---

@chayleaf 

I converted the `gradleScript` from `writeScript` to `writeShellApplication`, which not only looks better imo, but was also necessary in order to replace nsjail with bubblewrap.

I don't know how better bubblewrap is compared to nsjail, but it could be a viable alternative.

<details>

<summary> 0001-Rewrite-gradleScript.patch </summary>

```diff
diff --git a/pkgs/build-support/gradle/update-deps.nix b/pkgs/build-support/gradle/update-deps.nix
index 5d18092754e0..c76f8507ac57 100644
--- a/pkgs/build-support/gradle/update-deps.nix
+++ b/pkgs/build-support/gradle/update-deps.nix
@@ -3,8 +3,10 @@
 , stdenv
 , writeScript
 , writeShellScript
+, writeShellApplication
 , nix
 , gradle
+, coreutils
 , openssl
 , mitm-cache
 , jdk
@@ -44,49 +46,60 @@
 
 let
   mitmCachePortBash =
-    if mitmCachePort != null then "${toString mitmCachePort}"
-    else "$(${python3.pkgs.ephemeral-port-reserve}/bin/ephemeral-port-reserve 127.0.0.1)";
-  keep = [ "MITM_CACHE_PORT" "MITM_CACHE_KEYSTORE" "MITM_CACHE_KS_PWD" "MITM_CACHE_CA" ];
-  availablePackages' = map (drv: ''(builtins.storePath "${drv}")'') availablePackages;
-  gradleScript = writeScript "gradle-commands.sh" ''
-    #!/usr/bin/env nix-shell
-    #!nix-shell --pure ${lib.concatMapStringsSep " " (x: "--keep ${x}") keep} -i bash -p coreutils ${lib.concatStringsSep " " availablePackages'}
-    set -eo pipefail
-    source="$1"
-    gradleArgs=(
-      ${lib.escapeShellArgs gradleFlags}
-      --no-daemon
-      --console plain
-      --init-script ${./init-deps.gradle}
-      -Dhttp.proxyHost="127.0.0.1"
-      -Dhttps.proxyHost="127.0.0.1"
-      -Dhttp.proxyPort="$MITM_CACHE_PORT"
-      -Dhttps.proxyPort="$MITM_CACHE_PORT"
-      -Djavax.net.ssl.trustStore="$MITM_CACHE_KEYSTORE"
-      -Djavax.net.ssl.trustStorePassword="$MITM_CACHE_KS_PWD"
-    )
-    unset MITM_CACHE_KEYSTORE
-    unset MITM_CACHE_KS_PWD
-    function gradle() {
-      ${gradle}/bin/gradle $gradleFlags "''${gradleFlagsArray[@]}" "''${gradleArgs[@]}" "$@"
-    }
-    export GRADLE_USER_HOME="$(mktemp -d)"
-    export TERM=dumb
-    trap "rm -rf '$GRADLE_USER_HOME'" SIGINT SIGTERM ERR EXIT
-    export http_proxy="127.0.0.1:$MITM_CACHE_PORT"
-    export https_proxy="127.0.0.1:$MITM_CACHE_PORT"
-    export NIX_SSL_CERT_FILE="$MITM_CACHE_CA"
-    export SSL_CERT_FILE="$MITM_CACHE_CA"
-    SRC="$(mktemp -d)"
-    trap "rm -rf '$SRC'" SIGINT SIGTERM ERR EXIT
-    cp -r "$source/src" "$SRC/"
-    chmod -R +w "$SRC/src"
-    cd "$SRC/src"
-    source "$source/script"
-    ${preBuild}
-    ${gradleCommands}
-    ${postBuild}
-  '';
+    if mitmCachePort != null then
+      "${toString mitmCachePort}"
+    else
+      "$(${python3.pkgs.ephemeral-port-reserve}/bin/ephemeral-port-reserve 127.0.0.1)";
+  keep = [
+    "MITM_CACHE_PORT"
+    "MITM_CACHE_KEYSTORE"
+    "MITM_CACHE_KS_PWD"
+    "MITM_CACHE_CA"
+  ];
+  gradleScript = writeShellApplication {
+    name = "gradle-commands.sh";
+    runtimeInputs = availablePackages ++ [ coreutils ];
+    text = ''
+      set +o nounset
+      source="$1"
+      gradleArgs=(
+        ${lib.escapeShellArgs gradleFlags}
+        --no-daemon
+        --console plain
+        --init-script ${./init-deps.gradle}
+        -Dhttp.proxyHost="127.0.0.1"
+        -Dhttps.proxyHost="127.0.0.1"
+        -Dhttp.proxyPort="$MITM_CACHE_PORT"
+        -Dhttps.proxyPort="$MITM_CACHE_PORT"
+        -Djavax.net.ssl.trustStore="$MITM_CACHE_KEYSTORE"
+        -Djavax.net.ssl.trustStorePassword="$MITM_CACHE_KS_PWD"
+      )
+      unset MITM_CACHE_KEYSTORE
+      unset MITM_CACHE_KS_PWD
+      function gradle() {
+        # shellcheck disable=SC2154
+        ${gradle}/bin/gradle "''${gradleFlags}" "''${gradleFlagsArray[@]}" "''${gradleArgs[@]}" "$@"
+      }
+      # shellcheck disable=SC2155
+      export GRADLE_USER_HOME="$(mktemp -d)"
+      export TERM=dumb
+      trap 'rm -rf "$GRADLE_USER_HOME"' SIGINT SIGTERM ERR EXIT
+      export http_proxy="127.0.0.1:$MITM_CACHE_PORT"
+      export https_proxy="127.0.0.1:$MITM_CACHE_PORT"
+      export NIX_SSL_CERT_FILE="$MITM_CACHE_CA"
+      export SSL_CERT_FILE="$MITM_CACHE_CA"
+      SRC="$(mktemp -d)"
+      trap 'rm -rf "$SRC"' SIGINT SIGTERM ERR EXIT
+      cp -r "$source/src" "$SRC/"
+      chmod -R +w "$SRC/src"
+      cd "$SRC/src"
+      # shellcheck disable=SC1091
+      source "$source/script"
+      ${preBuild}
+      ${gradleCommands}
+      ${postBuild}
+    '';
+  };
 in
 writeShellScript "fetch-deps.sh" ''
   set -eo pipefail
-- 
2.45.1
```

</details>

<details>

<summary> 0002-Replace-nsjail-with-bubblewrap.patch </summary>

```diff
diff --git a/pkgs/build-support/gradle/update-deps.nix b/pkgs/build-support/gradle/update-deps.nix
index c76f8507ac57..0c1dea1d0673 100644
--- a/pkgs/build-support/gradle/update-deps.nix
+++ b/pkgs/build-support/gradle/update-deps.nix
@@ -10,7 +10,7 @@
 , openssl
 , mitm-cache
 , jdk
-, nsjail
+, bubblewrap
 , path
 , python3
 }:
@@ -50,12 +50,6 @@ let
       "${toString mitmCachePort}"
     else
       "$(${python3.pkgs.ephemeral-port-reserve}/bin/ephemeral-port-reserve 127.0.0.1)";
-  keep = [
-    "MITM_CACHE_PORT"
-    "MITM_CACHE_KEYSTORE"
-    "MITM_CACHE_KS_PWD"
-    "MITM_CACHE_CA"
-  ];
   gradleScript = writeShellApplication {
     name = "gradle-commands.sh";
     runtimeInputs = availablePackages ++ [ coreutils ];
@@ -149,13 +143,18 @@ writeShellScript "fetch-deps.sh" ''
   export MITM_CACHE_KEYSTORE="$PWD/keystore"
   popd
   ${preBuildNoSandbox}
-  # nsjail isn't necessary, it's only used to prevent messy build scripts from touching ~
-  ${nsjail}/bin/nsjail \
-    -N --keep_caps --chroot / --disable_rlimits \
-    -T /home -T /root -B /tmp -R ${path} \
-    ${lib.escapeShellArgs (map (x: "-E${x}") keep)} \
-    -E NIX_PATH=nixpkgs=${path} \
-    ${gradleScript} "$source"
+  # bubblewrap isn't necessary, it's only used to prevent messy build scripts from touching ~
+  ${bubblewrap}/bin/bwrap \
+    --ro-bind /nix /nix \
+    --bind-try /tmp /tmp \
+    --share-net \
+    --clearenv \
+    --setenv NIX_PATH nixpkgs="${toString ./.}" \
+    --setenv MITM_CACHE_PORT $MITM_CACHE_PORT \
+    --setenv MITM_CACHE_KEYSTORE $MITM_CACHE_KEYSTORE \
+    --setenv MITM_CACHE_KS_PWD $MITM_CACHE_KS_PWD \
+    --setenv MITM_CACHE_CA $MITM_CACHE_CA \
+    "${gradleScript}/bin/gradle-commands.sh" "$source"
   ${postBuildNoSandbox}
   kill -SIGINT "$MITM_CACHE_PID"
   for i in {0..20}; do
-- 
2.45.1
```

</details>

That said, I don't know if it's necessary to bind `/home` and `/root` in the first place as we're mainly working with things in `/nix/store` and `/tmp`.

---

your solution doesn't work as it can't add packages such as `glib.dev` for Gradle to use. Either way, this particular issue is resolved as i've added the mount now.